File: 백준/Platinum/1395. 스위치/스위치.py
import sys
input = lambda: sys.stdin.readline().strip()
n, m = map(int, input().split())

class Node:
    def __init__(self, start, end, left=None, right=None):
        self.range = (start, end)
        self.data = 0  # 해당 구간에서 켜진 전구의 개수
        self.lazy = 0  # lazy propagation을 위한 값
        self.left_child = left
        self.right_child = right

class Seg:
    def __init__(self):
        self.root = self.create(0, n - 1)

    def create(self, start, end):
        if start == end:
            return Node(start, end)
        mid = (start + end) // 2
        return Node(start, end,
            left=self.create(start, mid),
            right=self.create(mid + 1, end)
        )

    def propagate(self, cur: Node):
        """ Lazy propagation을 수행하여 현재 노드의 lazy 값을 반영하고 자식에게 전달 """
        start, end = cur.range
        if cur.lazy:
            cur.data = (end - start + 1) - cur.data  # 토글 연산
            if cur.left_child:
                cur.left_child.lazy ^= 1  # XOR 연산으로 토글 (0->1, 1->0)
                cur.right_child.lazy ^= 1
            cur.lazy = 0  # 현재 노드의 lazy 값을 초기화

    def modify(self, cur: Node, s, e):
        """ 범위 [s, e]의 전구 상태를 토글 (0->1, 1->0) """
        start, end = cur.range
        self.propagate(cur)  # Lazy 값을 현재 노드에 적용 후 진행

        if s > end or e < start:  # 범위를 벗어난 경우
            return
        if s <= start and end <= e:  # 범위 전체가 포함되는 경우
            cur.lazy ^= 1  # Lazy 값 반영
            self.propagate(cur)  # 현재 노드에 즉시 반영
            return

        # 구간을 나누어 좌우 자식에 대해 처리
        self.modify(cur.left_child, s, e)
        self.modify(cur.right_child, s, e)

        # 자식 노드들의 데이터를 합산하여 갱신
        cur.data = cur.left_child.data + cur.right_child.data

    def search(self, cur: Node, s, e):
        """ 범위 [s, e]의 전구 개수를 반환 """
        start, end = cur.range
        self.propagate(cur)  # Lazy 값을 현재 노드에 적용

        if s > end or e < start:  # 범위를 벗어난 경우
            return 0
        if s <= start and end <= e:  # 범위 전체가 포함되는 경우
            return cur.data

        # 부분적으로 겹치는 경우 좌우 자식 탐색 후 합산
        return self.search(cur.left_child, s, e) + self.search(cur.right_child, s, e)

segtree = Seg()

for _ in range(m):
    o, s, t = map(int, input().split())
    if o == 0:
        segtree.modify(segtree.root, s - 1, t - 1)
    else:
        print(segtree.search(segtree.root, s - 1, t - 1))

File: 백준/Platinum/1395. 스위치/스위치.java
import java.io.*;
import java.util.*;

class Node {
    int start, end, data, lazy;
    Node left, right;

    Node(int start, int end) {
        this.start = start;
        this.end = end;
        this.data = 0;
        this.lazy = 0;
    }
}

class Seg {
    Node root;

    Seg(int n) {
        root = create(0, n - 1);
    }

    private Node create(int start, int end) {
        if (start == end) {
            return new Node(start, end);
        }
        int mid = (start + end) / 2;
        Node node = new Node(start, end);
        node.left = create(start, mid);
        node.right = create(mid + 1, end);
        return node;
    }

    private void propagate(Node node) {
        if (node.lazy == 0) return;
        node.data = (node.end - node.start + 1) - node.data;
        if (node.left != null) {
            node.left.lazy ^= 1;
            node.right.lazy ^= 1;
        }
        node.lazy = 0;
    }

    void modify(Node node, int s, int e) {
        propagate(node);
        if (s > node.end || e < node.start) return;
        if (s <= node.start && node.end <= e) {
            node.lazy ^= 1;
            propagate(node);
            return;
        }
        modify(node.left, s, e);
        modify(node.right, s, e);
        node.data = node.left.data + node.right.data;
    }

    int search(Node node, int s, int e) {
        propagate(node);
        if (s > node.end || e < node.start) return 0;
        if (s <= node.start && node.end <= e) return node.data;
        return search(node.left, s, e) + search(node.right, s, e);
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader buffed = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer token = new StringTokenizer(buffed.readLine());
        int n = Integer.parseInt(token.nextToken());
        int m = Integer.parseInt(token.nextToken());

        Seg segtree = new Seg(n);
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < m; i++) {
            token = new StringTokenizer(buffed.readLine());
            int o = Integer.parseInt(token.nextToken());
            int s = Integer.parseInt(token.nextToken()) - 1;
            int t = Integer.parseInt(token.nextToken()) - 1;

            if (o == 0) {
                segtree.modify(segtree.root, s, t);
            } else {
                sb.append(segtree.search(segtree.root, s, t)).append("\n");
            }
        }
        System.out.print(sb);
    }
}